### ISSUE TITLE
Support investment transactions

### ISSUE THREAD

**Issue by JeffreyFalgout** [reubano/csv2ofx#32]
My HSA provider doesn't support Quicken export anymore. It does support Excel export, however. I was hoping to use this tool to bridge the gap, but unfortunately, this tool doesn't seem to support investment transactions (<INVSTMTMSGSETV1> in OFX parlance).

**Comment by jaraco** [comment]
Investment transactions are harder (impossible?) because each transaction must be associated with a security and each security must be uniquely identified (at the server). Unless your CSV provides a Unique ID (CUSIP) for each security, or unless the library can somehow solicit that information out-of-band, it may not be possible to generate a viable OFX file for investment transactions.

I'm in a similar situation where my payroll provider (Paychex) is only providing broken OFX downloads, so I'm writing a bespoke routine to convert the CSV downloads to OFX, but I'm having to parse out the securities list to resolve those unique ids.

**Comment by reubano** [comment]
Theres an investment csv in the data folder and an accompanying test for it as well. Let me know if that works for you or not.

**Comment by reubano** [comment]
The csv should provide unique security ids. And your accounting program should then be able to map ids to cusips or whatever else you need. Gnucash works that way at least.

**Comment by pjain** [comment]
Thanks for this. I would like to use it in a custom script as well as from the command line to import investment transactions. I have unique codes for the securities that would likely be imported but I'm not clear on what the source CSV for importing the transactions and the securities should look like. Can you give an example of what the CSV file should look like to import the securities as well as create the corresponding transactions?
Thanks!

**Comment by jaraco** [comment]
For what it's worth - [here's the routine I used](https://github.com/jaraco/jaraco.financial/blob/master/jaraco/financial/paychex.py) to create an OFX of the security transactions for PayChex. As you can see, it relies on the broken OFX download to get security metadata and header info that's not present in the CSV. It's not a generalizable approach, but it does start to reveal the interfaces that would be necessary to support a generalized approach.

**Comment by reubano** [comment]
so it appears that while I've [added investment support](https://github.com/reubano/csv2ofx/commit/45d8e0f15ce0b1f8b7a848acd416ae71378460eb), I didn't add a test, nor did I supply a proper csv file. @jaraco is there any chance you could take your csv file and modify your python script to use the investment feature?

**Comment by jaraco** [comment]
Yes, perhaps. I'll take a stab at it at some point.

**Comment by jaraco** [comment]
Here are the first two lines of an export from my paychex account:

```
Date, Transaction, InvestmentName, Ticker, Amount, Price, Shares
01/09/2018,Mid-Atlantic Trustee Fee,American Funds Washington Mutual Investors R6,RWMGX,-0.12,46.810000,-0.0026,
```

As you can see, it doesn't have the CUSIP... but I thought maybe I could add a more complex mapping that would load the CUSIP from another source. So I started attempting to create the mapping and running the package, but I get an error:

```
$ .tox/python/bin/python -m csv2ofx.main -m paychex ~/Downloads/EXPORT.csv export.ofx   
'<' not supported between instances of 'NoneType' and 'NoneType'
No data to write. '<' not supported between instances of 'NoneType' and 'NoneType'. Try again with `-c` option.
```

And here's the debug output:

```
$ .tox/python/bin/python -m csv2ofx.main -m paychex ~/Downloads/EXPORT.csv export.ofx --debug
{'account_type': None,
 'chunksize': 16384,
 'collapse': None,
 'custom': None,
 'debug': True,
 'dest': 'export.ofx',
 'encoding': 'utf-8',
 'end': '2018-02-11 13:16:41.057679',
 'language': 'ENG',
 'list_mappings': False,
 'mapping': 'paychex',
 'overwrite': False,
 'qif': False,
 'server_date': None,
 'source': '/Users/jaraco/Downloads/EXPORT.csv',
 'start': None,
 'verbose': False,
 'version': False}
```

Here's the mapping I used:

```
mapping = {
    'has_header': True,
    'is_split': False,
    'currency': 'USD',
    'delimiter': ',',
    'date': itemgetter('Date'),
    'amount': itemgetter('Amount'),
    'price': itemgetter('Price'),
    'shares': itemgetter('Shares'),
    'investment': itemgetter('InvestmentName'),
    'ticker': itemgetter('Ticker'),
    'desc': itemgetter('Transaction'),
}
```

Can you suggest how I could further troubleshoot - and what things might be missing from the mapping file or usage that I could supply to get some output?

**Comment by reubano** [comment]
After removing extra spaces in your csv header, and pushing a few fixes, I get the following:

```xml
DATA:OFXSGML
ENCODING:UTF-8
<OFX>
	<SIGNONMSGSRSV1>
		<SONRS>
			<STATUS>
				<CODE>0</CODE>
				<SEVERITY>INFO</SEVERITY>
			</STATUS>
			<DTSERVER>20180627143708</DTSERVER>
			<LANGUAGE>ENG</LANGUAGE>
		</SONRS>
	</SIGNONMSGSRSV1>
	<BANKMSGSRSV1>
		<STMTTRNRS>
			<TRNUID></TRNUID>
			<STATUS>
				<CODE>0</CODE>
				<SEVERITY>INFO</SEVERITY>
			</STATUS>
			<STMTRS>
				<CURDEF>USD</CURDEF>
				<BANKACCTFROM>
					<BANKID>382b0f5185773fa0f67a8ed8056c7759</BANKID>
					<ACCTID>382b0f5185773fa0f67a8ed8056c7759</ACCTID>
					<ACCTTYPE>CHECKING</ACCTTYPE>
				</BANKACCTFROM>
				<BANKTRANLIST>
					<DTSTART>19700101</DTSTART>
					<DTEND>20180627</DTEND>
					<STMTTRN>
						<TRNTYPE>DEBIT</TRNTYPE>
						<DTPOSTED>20180109000000</DTPOSTED>
						<TRNAMT>0.12</TRNAMT>
						<FITID>fa5d3eb2d66fbd854953d96cd49f41e7</FITID>
						<MEMO>Mid-Atlantic Trustee Fee</MEMO>
					</STMTTRN>
				</BANKTRANLIST>
			</STMTRS>
		</STMTTRNRS>
	</BANKMSGSRSV1>
</OFX>
```

**Comment by jwittkoski** [comment]
@lifekaizen Do you have a branch with your investment changes?

**Comment by lifekaizen** [comment]
> @lifekaizen Do you have a branch with your investment changes?

Yes, here's the branch: https://github.com/lifekaizen/csv2ofx/tree/investments

**Comment by reubano** [comment]
CR #37

**Comment by tim-rohrer** [comment]
> > @lifekaizen Do you have a branch with your investment changes?
> 
> Yes, here's the branch: https://github.com/lifekaizen/csv2ofx/tree/investments

Hey @enxyz, I'd like to experiment with this. I'm by no means a git or Python wizard, but I'll try forking your repo. I do see it is behind @reubano's so perhaps you can bring it forward? I'm not sure I'll be able to.

Then, perhaps together we can get a PR suitable for this repo?
